src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs

```diff
@@ -15240,6 +15252,361 @@ public C()
+    {
+        return f();
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib(source, references: s_valueTupleRefs);
+            comp.VerifyDiagnostics(
+                // (12,46): warning CS8123: The tuple element name 'b' is ignored because a different name is specified by the assignment target.
+                //                             return (a: 1, b: 2);
+                Diagnostic(ErrorCode.WRN_TupleLiteralNameMismatch, "2").WithArguments("b").WithLocation(12, 46),
+                // (18,50): warning CS8123: The tuple element name 'c' is ignored because a different name is specified by the assignment target.
+                //                                 return (a: d, c: 3);
+                Diagnostic(ErrorCode.WRN_TupleLiteralNameMismatch, "3").WithArguments("c").WithLocation(18, 50),
+                // (22,50): warning CS8123: The tuple element name 'd' is ignored because a different name is specified by the assignment target.
+                //                                 return (a: d, d: 4);
+                Diagnostic(ErrorCode.WRN_TupleLiteralNameMismatch, "4").WithArguments("d").WithLocation(22, 50)
+                );
+
+            var tree = comp.SyntaxTrees.First();
+            var model = comp.GetSemanticModel(tree);
+            var nodes = tree.GetCompilationUnitRoot().DescendantNodes();
+            var x1 = model.GetDeclaredSymbol(nodes.OfType<VariableDeclaratorSyntax>().First());
+            Assert.Equal("(dynamic a, System.Int32) x1", x1.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void LambdaTypeInferenceFails()
+        {
+            var source = @"
+public class C
+{
+    public void M()
+    {
+        var x1 = M2(() =>
+                    {
+                        bool flag = true;
+                        dynamic d = null;
+                        if (flag)
+                        {
+                            return (a: 1, b: d);
+                        }
+                        else
+                        {
+                            if (flag)
+                            {
+                                return (a: d, c: 3);
+                            }
+                            else
+                            {
+                                return (a: d, d: 4);
+                            }
+                        }
+                    });
+    }
+    public T M2<T>(System.Func<T> f)
+    {
+        return f();
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib(source, references: s_valueTupleRefs);
+            comp.VerifyDiagnostics(
+                // (6,18): error CS0411: The type arguments for method 'C.M2<T>(Func<T>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //         var x1 = M2(() =>
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "M2").WithArguments("C.M2<T>(System.Func<T>)").WithLocation(6, 18)
```

Comment:
So the current logic for finding the best type only merges types when they can be converted implicitly to each other. But currently there is no implicit conversion from `(System.Int32 a, dynamic b)` to `(dynamic a, System.Int32 c)` or back.
@VSadov That seems surprising. Is that what you expect?